pallets/inactivity-tracking/src/lib.rs

```diff
@@ -0,0 +1,258 @@
+        type MaxCollatorsPerSession: Get<u32>;
+
+        /// The maximum amount of container chains that can be stored
+        #[pallet::constant]
+        type MaxContainerChains: Get<u32>;
+
+        /// Helper that returns the current session index.
+        type CurrentSessionIndex: GetSessionIndex<SessionIndex>;
+
+        /// Helper that returns the block author for the orchestrator chain (if it exists)
+        type GetSelfChainBlockAuthor: MaybeSelfChainBlockAuthor<Self::CollatorId>;
+
+        /// Helper that fetches the latest set of container chains and their collators
+        type ContainerChainsFetcher: GetContainerChainsWithCollators<Self::CollatorId>;
+
+        /// Helper that allows to check if a new session will start in the next block
+        type SessionEndChecker: GetRandomnessForNextBlock<BlockNumberFor<Self>>;
+    }
+
+    /// A list of double map of inactive collators for a session
+    #[pallet::storage]
+    pub type ActiveCollators<T: Config> = StorageMap<
+        _,
+        Twox64Concat,
+        SessionIndex,
+        BoundedVec<T::CollatorId, T::MaxCollatorsPerSession>,
+        ValueQuery,
+    >;
+
+    /// A list of inactive collators for a session. Repopulated at the start of every session
+    #[pallet::storage]
+    pub type ActiveCollatorsForCurrentSession<T: Config> =
+        StorageValue<_, BoundedVec<T::CollatorId, T::MaxCollatorsPerSession>, ValueQuery>;
+
+    /// A list of inactive container chains for a session. Repopulated at the start of every session
+    #[pallet::storage]
+    pub type ActiveContainerChainsForCurrentSession<T: Config> =
+        StorageValue<_, BoundedVec<ParaId, T::MaxInactiveSessions>, ValueQuery>;
+
+    /// The last session index for which the inactive collators have not been processed
+    #[pallet::storage]
+    pub type LastUnprocessedSession<T: Config> = StorageValue<_, SessionIndex, ValueQuery>;
+
+    #[pallet::event]
+    pub enum Event<T: Config> {}
+
+    #[pallet::error]
+    pub enum Error<T> {
+        MaxCollatorsPerSessionReached,
+        MaxContainerChainsReached,
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {}
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
+            let mut total_weight = T::DbWeight::get().reads_writes(1, 0);
+
+            // Process the orchestrator chain block author (if it exists)
+            if let Some(orchestrator_chain_author) = T::GetSelfChainBlockAuthor::get_block_author()
+            {
+                total_weight += Self::on_author_noted(orchestrator_chain_author);
+            }
+
+            // Update inactive collator records only after a session has ended
+            let current_session = T::CurrentSessionIndex::session_index();
+            let current_unprocessed_session = <LastUnprocessedSession<T>>::get();
+            if current_unprocessed_session < current_session {
```

Comment:
A way to avoid 2 reads every block is doing this inside `apply_own_session`. However I dont dislike this

Reply:
or inside the function `before_session_ending`

Reply:
Good point...  Using `before_session_ending ` should remove the need for storing `LastUnprocessedSession` session too... I will add the `SessionHandler` hook for dealing with  chain inactivity first and update this afterwards

Reply:
Changes will be applied in https://github.com/moondance-labs/tanssi/tree/aleks-collator-inactivity-tracking